// prehistoric/lib/commonjs/history.js
"use strict";
var most_1 = require("most");
var most_subject_1 = require("most-subject");
var BrowserSource_1 = require("./BrowserSource");
var ServerSource_1 = require("./ServerSource");
function createHistory() {
    var holdSource = new most_subject_1.HoldSubjectSource(most_1.never().source, 1);
    var inBrowser = currentlyInBrowser();
    var historySource = inBrowser ? new BrowserSource_1.BrowserSource() : new ServerSource_1.ServerSource();
    if (inBrowser) {
        window.addEventListener('popstate', function (ev) {
            ev.preventDefault();
            event(historySource.getCurrentLocation());
        });
    }
    var history = new most_1.Stream(holdSource).startWith(historySource.go(0));
    function event(location) {
        most_1.defaultScheduler.asap(most_1.PropagateTask.event(location, holdSource));
    }
    function push(path, state) {
        if (state === void 0) { state = {}; }
        event(historySource.push(path, state));
    }
    function replace(path, state) {
        if (state === void 0) { state = {}; }
        event(historySource.replace(path, state));
    }
    function go(amount) {
        if (inBrowser)
            return historySource.go(amount);
        event(historySource.go(amount));
    }
    return { push: push, replace: replace, go: go, history: history };
}
exports.createHistory = createHistory;
function currentlyInBrowser() {
    try {
        return window && window.history;
    }
    catch (e) {
        return false;
    }
}
