// hucklebuck/src/lib/analytics/destinations/google.js
import { bus } from 'partybus';

const embed = (config) => {
  let script = document.createElement("script");
  const gte = `(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
})(window,document,'script','dataLayer','${config.googletag_api_key}');`;
  script.innerHTML = gte;
  document.head.appendChild(script);
};

const data_layer_init = (config, actor) => {
  let script = document.createElement("script");
  script.innerHTML = `
  dataLayer = [{
    'activity_id': '${config.full_url}',
    'activity_name': '${config.activity_name}',
    'fullName': '${actor.name}',
    'email': '${actor.email}'
  }];`;
  document.head.appendChild(script);
};

const event_track = (name, data) => {
  data.event = name;
  console.log("Google Tag track:", name, data);
  window.dataLayer.push(data);
};

const fake_push = (name, data) => {
  console.log("Google Tag track (faked):", name, data);
}

const googletag = (setup_tor_real, config, actor) => {
  let google_push = setup_tor_real ? event_track : fake_push; 
  if (setup_tor_real) {
    data_layer_init(config, actor);
    embed(config);
  }
  const opening_time = Date.now();
  // https://developers.google.com/tag-manager/devguide
  // Google Tag Embed must come AFTER the data layer script
  // according to the developer guides

  
  bus.on('hucklebucked', (slug_or_num) => {
    var tracking_data = {
      activity: config.activity_name,
      activity_id: config.full_url,
      section: slug_or_num,
      timestamp: Date.now(),
      elapsed: Math.round(Date.now() - opening_time / 1000),
      name: actor.name,
      email: actor.email,
    };
    google_push("navigateSection", tracking_data);
  });

  bus.on('activity::progress::mark_as_completed', (time) => {
    var tracking_data = {
      activity: config.activity_name,
      activity_id: config.full_url,
      timestamp: time,
      name: actor.name,
      email: actor.email,
    };
    google_push("completedActivity", tracking_data);
  });
}

export default googletag;
