// markdown-brunch/lib/index.js
// Generated by CoffeeScript 1.6.3
var MarkdownCompiler, hljs, marked, umd;

marked = require('marked');

hljs = require('highlight.js');

umd = require('umd-wrapper');

module.exports = MarkdownCompiler = (function() {
  MarkdownCompiler.prototype.brunchPlugin = true;

  MarkdownCompiler.prototype.type = 'template';

  MarkdownCompiler.prototype.extension = 'md';

  MarkdownCompiler.prototype.pattern = /(\.(markdown|mdown|mkdn|md|mkd|mdwn|mdtxt|mdtext|text))$/;

  function MarkdownCompiler(config) {
    var languages, options, _ref;
    languages = Object.keys(hljs.LANGUAGES);
    options = Object.create((_ref = config.marked) != null ? _ref : null);
    if (!('highlight' in options)) {
      options.highlight = function(code, lang) {
        if (lang === 'auto') {
          return hljs.highlightAuto(code).value;
        } else if (languages.indexOf(lang) !== -1) {
          return hljs.highlight(lang, code).value;
        }
      };
    }
    marked.setOptions(options);
  }

  MarkdownCompiler.prototype.compile = function(data, path, callback) {
    var err, error, result;
    try {
      return result = umd(JSON.stringify(marked(data)));
    } catch (_error) {
      err = _error;
      return error = err;
    } finally {
      callback(error, result);
    }
  };

  return MarkdownCompiler;

})();
