// hucklebuck/gulptasks/indexhtml.js
const gulp = require('gulp');
const postcss = require('gulp-postcss');
const htmlreplace = require('gulp-html-replace');
const htmlmin = require('gulp-htmlmin');

const processors = [
  require('postcss-import')(),
  require('postcss-url')(),
  require('postcss-cssnext')(),
  require('cssnano')({ autoprefixer: false }),
  require('postcss-browser-reporter')(),
  require('postcss-reporter')()
];

const indexhtml = function(config) {
  return function() {
    // This will use the CSS processors above to compile, compress and
    // inject the CSS into the HTML head.
    // Also it will inject other parts of the HTML into the document
    // that are easier to manage as separate pieces
    // Lastly it will compress the HTML file entirely.
    return gulp
      .src('./src/index.html')
      .pipe(
        htmlreplace({
          css: {
            src: gulp.src('./src/css/index.css').pipe(postcss(processors)),
            tpl: '<style>%s</style>'
          },
          title: {
            src: [
              [
                config.activity_name,
                config.activity_name,
                config.activity_name
              ]
            ],
            tpl: '<title>%s</title><meta name="description" content="%s"><meta name="apple-mobile-web-app-title" content="%s">'
          },
          author: {
            src: config.author,
            tpl: '<meta name="author" content="%s">'
          },
          navigation: { src: gulp.src('./src/ui/navigation.html'), tpl: '' },
          content: {
            src: gulp.src(`./src/sections/${config.first_section}.html`),
            tpl: ''
          },
          footerjs: { src: gulp.src('./footerjs.html'), tpl: '' },
          footer: { src: gulp.src('./src/ui/footer.html'), tpl: '' },
          modals: { src: gulp.src('./src/ui/modals/*.html'), tpl: '' }
        })
      )
      .pipe(htmlmin({ collapseWhitespace: true }))
      .pipe(gulp.dest(config.distpath));
  };
};

module.exports = indexhtml;
